src/Magnum/LibOvrIntegration/Compositor.cpp

```diff
@@ -0,0 +1,125 @@
+    DEALINGS IN THE SOFTWARE.
+*/
+
+#include "Magnum/LibOVRIntegration/Compositor.h"
+
+#include "Magnum/LibOVRIntegration/Conversion.h"
+#include "Magnum/LibOVRIntegration/Hmd.h"
+
+#include <OVR_CAPI_GL.h>
+
+namespace Magnum { namespace LibOvrIntegration {
+
+Layer::Layer(const LayerType type): _layer(), _TYPE(type) {
+    _layer.Header.Type = ovrLayerType(Corrade::Containers::EnumSet<LayerType>::UnderlyingType(_TYPE));
+    _layer.Header.Flags = ovrLayerFlag_TextureOriginAtBottomLeft;
+}
+
+Layer& Layer::setEnabled(bool enabled) {
+    if(enabled) {
+        _layer.Header.Type = ovrLayerType(Corrade::Containers::EnumSet<LayerType>::UnderlyingType(_TYPE));
+    } else {
+        _layer.Header.Type = ovrLayerType_Disabled;
+    }
+    return *this;
+}
+
+//----------------------------------------------------------------
+
+LayerEyeFov::LayerEyeFov(): Layer(LayerType::EyeFov) {
+}
+
+LayerEyeFov& LayerEyeFov::setColorTexture(const int eye, SwapTextureSet& textureSet) {
+    _layer.EyeFov.ColorTexture[eye] = &textureSet.getOvrSwapTextureSet();
+
+    return *this;
+}
+
+LayerEyeFov& LayerEyeFov::setViewport(const int eye, const Range2Di& viewport) {
+    _layer.EyeFov.Viewport[eye] = ovrRecti(viewport);
+
+    return *this;
+}
+
+LayerEyeFov& LayerEyeFov::setRenderPoses(const Hmd& hmd) {
+    _layer.EyeFov.RenderPose[0] = hmd._ovrPoses[0];
+    _layer.EyeFov.RenderPose[1] = hmd._ovrPoses[1];
+
+    return *this;
+}
+
+LayerEyeFov& LayerEyeFov::setFov(const Hmd& hmd) {
+    _layer.EyeFov.Fov[0] = hmd._hmd->DefaultEyeFov[0];
+    _layer.EyeFov.Fov[1] = hmd._hmd->DefaultEyeFov[1];
+
+    return *this;
+}
+
+//----------------------------------------------------------------
+
+Compositor::Compositor(): _layers(), _wrappedLayers(), _layersCount(0) {
+}
+
+typedef Corrade::Containers::EnumSet<LayerType> LayerTypes;
+
+Layer& Compositor::addLayer(LayerType type) {
+    switch (type) {
+        case LayerType::Direct:
+            CORRADE_ASSERT(true, "LayerType::Direct is not implemented.", addLayerEyeFov());
+            break;
+        case LayerType::EyeFov:
+            return addLayerEyeFov();
+        case LayerType::EyeFovDepth:
+            CORRADE_ASSERT(true, "LayerType::EyeFovDepth is not implemented.", addLayerEyeFov());
+            break;
+        case LayerType::QuadHeadLocked:
+            CORRADE_ASSERT(true, "LayerType::QuadHeadLocked is not implemented.", addLayerEyeFov());
+            break;
+        case LayerType::QuadInWorld:
+            CORRADE_ASSERT(true, "LayerType::QuadInWorld is not implemented.", addLayerEyeFov());
+            break;
+        default:
+            CORRADE_ASSERT_UNREACHABLE();
+    }
+    CORRADE_ASSERT_UNREACHABLE();
+}
+
+LayerEyeFov& Compositor::addLayerEyeFov() {
+    std::unique_ptr<Layer> wrapped(new LayerEyeFov());
+    _wrappedLayers.push_back(std::move(wrapped));
+    _layers.push_back(&wrapped->layerHeader());
+
+    return *static_cast<LayerEyeFov*>(_wrappedLayers.back().get());
```

Comment:
Just swapping the two `push_back()` calls should do it, no? :) As far as I can tell, tests for this would require active GL context. I'll push the needed changes for `AbstractOpenGLTester` this evening, hopefully.

Oh, looking above, the `CORRADE_ASSERT(true, ...)` calls should probably be `false`, otherwise they won't do anything. I'd remove the `default:` case as well, because then some compilers would be able to issue a warning when there's some unhandled enum value (very valuable warning, in my opinion). ... um, and, the `_TYPE` member should be rather `_type`, underscore followed by uppercase letter is reserved by the c++ standard.

Sorry, got carried away again :)